### PR TITLE
Feature/twitter spaces integration

### PR DESCRIPTION
**Pull request by amaredeus**
2. Add twitter spaces integration (wip)
   - refactor MongoDbUtils file
   - add POAP schedule error messaging
   - setup POAP claiming for failed delivery
   - remove deaf users for poap events
   - organize error message validations
   - add workflow env keys
   
   
   related: https://github.com/BanklessDAO/discord-bots-web/pull/5

### CODE REVIEW

File: src/app/commands/admin/Account.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import { LogUtils } from '../../utils/Log';
+import VerifyTwitter from '../../service/account/VerifyTwitter';
+import ServiceUtils from '../../utils/ServiceUtils';
+import discordServerIds from '../../service/constants/discordServerIds';
+
+export default class Account extends SlashCommand {
+	constructor(creator: SlashCreator) {
+		super(creator, {
+			name: 'account',
+			description: 'Manage your accounts integration.',
```
Comment (nonsensecodes) [inline src/app/commands/admin/Account.ts:11]: Can we do `Manage your account's integration.`?

File: src/app/commands/admin/Account.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		});
+	}
+	
+	async run(ctx: CommandContext): Promise<any> {
+		LogUtils.logCommandStart(ctx);
+		if (ctx.user.bot) return;
+		const { guildMember } = await ServiceUtils.getGuildAndMember(ctx);
+		try {
+			await VerifyTwitter(guildMember);
+			await ctx.send({ content: 'Twitter verification sent to your DM!' });
```
Comment (nonsensecodes) [inline src/app/commands/admin/Account.ts:48]: 'Let\'s continue your Twitter verification in your DMs.'

File: src/app/commands/admin/Account.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		} catch (e) {
+			LogUtils.logError('failed to verify user', e, guildMember.guild.id);
+			await ctx.send('Something is not working...');
```
Comment (nonsensecodes) [inline src/app/commands/admin/Account.ts:51]: 'I\'d say you broke this, but this is probably a me thing and not a you thing. I\'ll tell my devs to look into this.'

File: src/app/commands/poap/POAP.ts [hunk 4]
```diff
@@ -137,20 +194,25 @@ module.exports = class poap extends SlashCommand {
...
+				return;
 			default:
 				return ctx.send(`${ctx.user.mention} Please try again.`);
 			}
-			return this.handleCommandError(ctx, command);
+			this.handleCommandError(ctx, command);
+			return;
 		} catch (e) {
 			LogUtils.logError('failed to process POAP command', e);
+			return ctx.send('Sorry something is not working and our devs are looking into it.');
```
Comment (nonsensecodes) [inline src/app/commands/poap/POAP.ts:215]: 'Welp, something is definitely broken. I would blame you, but I know better. I\'ll let my devs know something is wrong.'

File: src/app/commands/poap/POAP.ts [hunk 5]
```diff
@@ -168,7 +230,7 @@ module.exports = class poap extends SlashCommand {
 				return ctx.send(e.message);
 			} else {
 				LogUtils.logError('failed to handle poap command', e);
-				return ctx.send('Sorry something is not working and our devs are looking into it.');
+				return ctx.send('Nothing to see here..');
```
Comment (nonsensecodes) [inline src/app/commands/poap/POAP.ts:233]: YES. 🙌

File: src/app/events/poap/AddUserForEvent.ts [hunk 6]
```diff
@@ -49,47 +50,52 @@ export const addUserToDb = async (
...
 		return;
 	}
-
-	// Check if user joined channel
-	if (newState.channelId === channel.id) {
-		await updateUserForPOAP(member, db, channel, true).catch(e => LogUtils.logError('failed to capture user joined for poap', e));
-	} else {
-		await updateUserForPOAP(member, db, channel, false).catch(e => LogUtils.logError('failed to capture user left for poap', e));
+	if (newState.deaf) {
+		await updateUserForPOAP(member, db, channel, false, true).catch(e => LogUtils.logError('failed to capture user joined for poap', e));
```
Comment (nonsensecodes) [inline src/app/events/poap/AddUserForEvent.ts:54]: 'failed to capture user who attempted to join for POAP'

File: src/app/service/constants/apiKeys.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+const apiKeys = Object.freeze({
```
Comment (nonsensecodes) [inline src/app/service/constants/apiKeys.ts:1]: Can we rename this file to `TwitterAPIKeys`? Or do you plan on packing this with all API keys in the future?

File: src/app/service/guest-pass/AddGuestPass.ts [hunk 8]
```diff
@@ -17,17 +17,17 @@ export default async (guestUser: GuildMember): Promise<any> => {
 	notifyUserOfGuestExpiration(guestUser);
 	removeGuestRoleOnExpiration(guestUser);
 	return guestUser.send({
-		content: `You have been granted guest access at Bankless DAO! Guest passes last for 
-		${process.env.DAO_GUEST_PASS_EXPIRATION_DAYS} days and will expire afterwards. To renew your guest pass, ask any 
-		Level 2 contributor to renew it, or post in #get-involved. In the future, we'll automate guest pass renewal based on 
-		your activity in the DAO!` })
-		.catch(e => LogUtils.logError('failed to send message to new guest', e));
+		content: 'You have been granted guest access at Bankless DAO! Guest passes last for '
```
Comment (nonsensecodes) [inline src/app/service/guest-pass/AddGuestPass.ts:20]: `Well look at you! Welcome to BanklessDAO! You've been given a guest pass by ${user}! Go give them a thank you! Your guest pass will last for ${time} at which time you should return to ${user} to renew your guest pass.`

File: src/app/service/poap/ClaimPOAP.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+import ServiceUtils from '../../utils/ServiceUtils';
+import POAPUtils from '../../utils/POAPUtils';
+
+const ClaimPOAP = async (ctx: CommandContext, guildMember: GuildMember, platform: string, code: string): Promise<any> => {
+	Log.debug(`starting claim with claimCode: ${code}`);
+	
+	POAPUtils.validateClaimCode(code);
+	
+	if (platform == constants.PLATFORM_TYPE_TWITTER) {
+		await ctx.send('Twitter platform is not supported at this time. Please reach out to community organizer for missing POAP');
```
Comment (nonsensecodes) [inline src/app/service/poap/ClaimPOAP.ts:17]: 'So, the Twitter platform isn't supported with what you\'re trying to do here. Reach out to community organizer to hook you up with your POAP.'

File: src/app/service/poap/ClaimPOAP.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	
+	const db: Db = await MongoDbUtils.connect(constants.DB_NAME_DEGEN);
+	const unclaimedParticipantsCollection: Collection<POAPUnclaimedParticipants> = await db.collection(constants.DB_COLLECTION_POAP_UNCLAIMED_PARTICIPANTS);
+	const unclaimedParticipants: Cursor<POAPUnclaimedParticipants> = await unclaimedParticipantsCollection.find({
+		discordUserId: guildMember.user.id,
+		claimCode: code,
+	});
+	
+	if (!await unclaimedParticipants.hasNext()) {
+		await ctx.send('Could not find any POAPs 🤷‍');
```
Comment (nonsensecodes) [inline src/app/service/poap/ClaimPOAP.ts:29]: 'Listen, I looked for the POAPs, but I failed to find them. I\'d apologize, but this is definitely not a me thing.'

File: src/app/service/poap/DistributePOAP.ts [hunk 11]
```diff
@@ -33,10 +35,15 @@ export default async (ctx: CommandContext, guildMember: GuildMember, event?: str
 		],
 		files: [{ name: 'failed_to_send_poaps.csv', attachment: failedPOAPsBuffer }],
 	});
+	if (failedPOAPsList.length <= 0) {
+		Log.debug('all poap successfully delivered');
```
Comment (nonsensecodes) [inline src/app/service/poap/DistributePOAP.ts:39]: 'All POAPs successfully delivered.'

File: src/app/service/poap/DistributePOAP.ts [hunk 12]
```diff
@@ -33,10 +35,15 @@ export default async (ctx: CommandContext, guildMember: GuildMember, event?: str
...
+		return;
+	}
+	await POAPUtils.setupFailedAttendeesDelivery(guildMember, failedPOAPsList, event, code, ctx);
 };
 
 export const askForParticipantsList = async (guildMember: GuildMember): Promise<POAPFileParticipant[]> => {
-	const message: Message = await guildMember.send({ content: 'Please upload participants.csv file.' });
+	const message: Message = await guildMember.send({ content: 'Please upload participants.csv file. POAPs will be distributed to these degens.' });
```
Comment (nonsensecodes) [inline src/app/service/poap/DistributePOAP.ts:46]: YES! 🙌

File: src/app/service/poap/EndPOAP.ts [hunk 13]
```diff
@@ -9,12 +8,17 @@ import POAPUtils, { FailedPOAPAttendee, POAPFileParticipant } from '../../utils/
 import { CommandContext } from 'slash-create';
 import Log from '../../utils/Log';
 import dayjs from 'dayjs';
+import MongoDbUtils from '../../utils/MongoDbUtils';
+import ServiceUtils from '../../utils/ServiceUtils';
 
-export default async (guildMember: GuildMember, ctx?: CommandContext): Promise<any> => {
-	const db: Db = await dbInstance.dbConnect(constants.DB_NAME_DEGEN);
+export default async (guildMember: GuildMember, code?: string, ctx?: CommandContext): Promise<any> => {
+	Log.debug('attempting to ending poap event');
```
Comment (nonsensecodes) [inline src/app/service/poap/EndPOAP.ts:15]: 'Attempting to end POAP event.'

File: src/app/service/poap/EndPOAP.ts [hunk 14]
```diff
@@ -58,6 +68,7 @@ export default async (guildMember: GuildMember, ctx?: CommandContext): Promise<a
 	const bufferFile = getBufferFromParticipants(listOfParticipants, channel);
 	const currentDate = (new Date()).toISOString();
 	const fileName = `${channel.guild.name}_${channel.name}_${listOfParticipants.length}_participants.csv`;
+	await guildMember.send({ content: 'The event has ended and here are the results.' });
```
Comment (nonsensecodes) [inline src/app/service/poap/EndPOAP.ts:71]: 'You did it! The event has ended\; here are the results:'

**Review by nonsensecodes — CHANGES_REQUESTED** [review]
Per our discussion regarding bot verbiage, I've made suggestions around what can be stated to give DEGEN a little more personality. Otherwise, killer fucking work, @amaredeus!

File: src/app/service/constants/apiKeys.ts (hunk 7 above)
Comment (amaredeus) [reply src/app/service/constants/apiKeys.ts:1]: Plan is to pack all api keys here going forward

File: src/app/commands/admin/Account.ts (hunk 1 above)
Comment (amaredeus) [reply src/app/commands/admin/Account.ts:11]: done

File: src/app/commands/admin/Account.ts (hunk 2 above)
Comment (amaredeus) [reply src/app/commands/admin/Account.ts:48]: done

File: src/app/service/poap/DistributePOAP.ts (hunk 11 above)
Comment (amaredeus) [reply src/app/service/poap/DistributePOAP.ts:39]: log messages are generally lower case and without punctuation to help with keeping the cost of storage to a minimum

File: src/app/service/poap/EndPOAP.ts (hunk 13 above)
Comment (amaredeus) [reply src/app/service/poap/EndPOAP.ts:15]: same here

File: src/app/service/poap/EndPOAP.ts (hunk 14 above)
Comment (amaredeus) [reply src/app/service/poap/EndPOAP.ts:71]: How about "Finally it's over..." XD

File: src/app/service/poap/DistributePOAP.ts (hunk 11 above)
Comment (nonsensecodes) [reply src/app/service/poap/DistributePOAP.ts:39]: Gotcha 👍

**Review by nonsensecodes — CHANGES_REQUESTED** [review]
Two spots that I didn’t see commentary or push back on. Just want to make sure they weren’t glossed over 👍

File: src/app/events/poap/AddUserForEvent.ts (hunk 6 above)
Comment (amaredeus) [reply src/app/events/poap/AddUserForEvent.ts:54]: ah this is also a log entry

File: src/app/service/guest-pass/AddGuestPass.ts (hunk 8 above)
Comment (amaredeus) [reply src/app/service/guest-pass/AddGuestPass.ts:20]: Ah yes the text here was requested from frog monkey, should we let him know we're replacing it?

File: src/app/service/poap/ClaimPOAP.ts (hunk 9 above)
Comment (amaredeus) [reply src/app/service/poap/ClaimPOAP.ts:17]: Oops missed this

File: src/app/service/guest-pass/AddGuestPass.ts (hunk 8 above)
Comment (amaredeus) [reply src/app/service/guest-pass/AddGuestPass.ts:20]: Hmm it doesn't seem discord captures who changes the role change when it happens.

Comment (nonsensecodes) [reply src/app/service/guest-pass/AddGuestPass.ts:20]: Oh that’s a bummer. Aight. To be continued.

**Review by nonsensecodes — APPROVED** [review]
Awesome! Thank you, @amaredeus 🙏

**Pull request merged**